### PR TITLE
LDrawLoader: 2 bugs and 7 improvements

### PR DESCRIPTION
**Pull request by yomboprime**
@gkjohnson

Related issue: None

Solved two bugs and made 7 improvements to LDrawLoader.
Bugs:
 - In parse(), missing call to applyMaterialsToMesh()
 - Missing assignment of userData.colorCode in conditional edge material.
 
Improvements:
 - Changed the order of search of files in the defaults directories, which improves load time.
 - Added 4 fields to userData: Author, fileName, type and colorCode.
 - If a color is not found, instead of giving error, give a warning and return a pink material.
 - Luminance parsing improvement (when the value is in the token, i.e. ```LUMINANCE15```)

I'm sorry I made it separated by commits but I lost that repository. It is all in one commit.

### CODE REVIEW

File: examples/jsm/loaders/LDrawLoader.js [hunk 1]
```diff
@@ -2080,8 +2102,24 @@ class LDrawLoader extends Loader {
 				material = loader.getMaterial( colorCode );
 				if ( material === null ) {
 
-					// otherwise throw an error if this is final opportunity to set the material
-					throw new Error( `LDrawLoader: Material properties for code ${ colorCode } not available.` );
+					// otherwise throw a warning if this is final opportunity to set the material
+					console.warn( `LDrawLoader: Material properties for code ${ colorCode } not available.` );
+
+					// And return the 'missing color' material
+					material = loader.missingColorMaterial.clone();
```
Comment (gkjohnson) [inline examples/jsm/loaders/LDrawLoader.js:2109]: Is this going to result in a ton of cloned materials if multiple parts use the same missing material color?

File: examples/jsm/loaders/LDrawLoader.js [hunk 2]
```diff
@@ -2370,6 +2409,35 @@ class LDrawLoader extends Loader {
 
 		return material;
 
+		function parseLuminance( token ) {
+
+			// Returns success
+
+			let lum;
+
+			if ( token.startsWith( "LUMINANCE" ) ) {
```
Comment (gkjohnson) [inline examples/jsm/loaders/LDrawLoader.js:2418]: nit: single qoutes

Comment (yomboprime) [reply examples/jsm/loaders/LDrawLoader.js:2418]: Thanks

**Comment by yomboprime** [comment]
Thanks!

File: examples/jsm/loaders/LDrawLoader.js [hunk 3]
```diff
@@ -2080,8 +2102,24 @@ class LDrawLoader extends Loader {
...
 
-					// otherwise throw an error if this is final opportunity to set the material
-					throw new Error( `LDrawLoader: Material properties for code ${ colorCode } not available.` );
+					// otherwise throw a warning if this is final opportunity to set the material
+					console.warn( `LDrawLoader: Material properties for code ${ colorCode } not available.` );
+
+					// And return the 'missing color' material
+					material = loader.missingColorMaterial;
+					material.name = "Missing material for color code '" + colorCode + "'";
+					material.userData.code = colorCode;
```
Comment (LeviPesin) [inline examples/jsm/loaders/LDrawLoader.js:2111]: So we return the same material but change its name and userData in every invocation?... If so, I think it should be cloned.

File: examples/jsm/loaders/LDrawLoader.js [hunk 4]
```diff
@@ -2080,8 +2102,24 @@ class LDrawLoader extends Loader {
...
+					const edgeMat = loader.missingEdgeColorMaterial.clone();
+					edgeMat.name = "Missing material for color code '" + colorCode + "' - Edge";
+					edgeMat.userData.code = colorCode;
+
+					const condEdgeMat = loader.missingConditionalEdgeColorMaterial.clone();
+					condEdgeMat.name = "Missing material for color code '" + colorCode + "' - Edge";
+					condEdgeMat.userData.code = colorCode;
+
+					material.userData.edgeMaterial = edgeMat;
+					material.userData.edgeMaterial.userData.conditionalEdgeMaterial = condEdgeMat;
```
Comment (LeviPesin) [inline examples/jsm/loaders/LDrawLoader.js:2122]: `material.userData.edgeMaterial` can be changed just to `edgeMat`.

File: examples/jsm/loaders/LDrawLoader.js [hunk 5]
```diff
@@ -1949,7 +1969,9 @@ class LDrawLoader extends Loader {
 			.parseModel( text, this.materialLibrary )
 			.then( group => {
 
+				this.applyMaterialsToMesh( group, MAIN_COLOUR_CODE, this.materialLibrary, true );
 				this.computeConstructionSteps( group );
+				group.userData.fileName = "";
```
Comment (LeviPesin) [inline examples/jsm/loaders/LDrawLoader.js:1974]: Here should also be single quotes.

File: examples/jsm/loaders/LDrawLoader.js (hunk 3 above)
Comment (yomboprime) [reply examples/jsm/loaders/LDrawLoader.js:2111]: You're right. @gkjohnson What do you think is better, to clone or to remove the name modification?

File: examples/jsm/loaders/LDrawLoader.js (hunk 4 above)
Comment (yomboprime) [reply examples/jsm/loaders/LDrawLoader.js:2122]: Thanks, done.

File: examples/jsm/loaders/LDrawLoader.js (hunk 5 above)
Comment (yomboprime) [reply examples/jsm/loaders/LDrawLoader.js:1974]: Thanks, I'll commit when the clone issue is decided.

**Comment by yomboprime** [comment]
Oh BTW I forgot to remove the cloning of the edge materials too.

File: examples/jsm/loaders/LDrawLoader.js (hunk 3 above)
Comment (yomboprime) [reply examples/jsm/loaders/LDrawLoader.js]: I've commited the change to generic missing material without clone(). Let me know if you agree.

Comment (LeviPesin) [reply examples/jsm/loaders/LDrawLoader.js]: 👍

Comment (gkjohnson) [reply examples/jsm/loaders/LDrawLoader.js]: I think this is good! The warning should be enough to let users know which information is missing.

File: examples/jsm/loaders/LDrawLoader.js [hunk 6]
```diff
@@ -2118,8 +2149,7 @@ class LDrawLoader extends Loader {
 
 	getMainEdgeMaterial() {
 
-		const mainMat = this.getMainMaterial();
-		return mainMat && mainMat.userData ? mainMat.userData.edgeMaterial : null;
+		return this.getMaterial( MAIN_EDGE_COLOUR_CODE );
```
Comment (gkjohnson) [inline examples/jsm/loaders/LDrawLoader.js:2152]: This previously used to return an "LineBasicMaterial" and now it returns a Mesh material - is this an intended change?

**Review by gkjohnson — APPROVED** [review]

**Comment by yomboprime** [comment]
@mrdoob this pr is ready to merge.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**